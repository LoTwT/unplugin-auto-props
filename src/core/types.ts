import type { createChecker } from "vue-component-meta"

export interface MappedRuntimeProp {
  /**
   * String | Number | Boolean | Array | Object | Date | Function | Symbol | Error
   */
  type: string
  required: boolean
}

export type MetaChecker = ReturnType<typeof createChecker>
