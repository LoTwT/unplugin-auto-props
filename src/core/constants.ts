export const EXTERNAL_META_PROPS = [
  "key",
  "ref",
  "ref_for",
  "ref_key",
  "class",
  "style",
]

export const COMPONENT_OPTION_KEYS = {
  PROPS: "props",
  EMITS: "emits",
} as const

export const DEFINE_COMPONENT = "defineComponent"
