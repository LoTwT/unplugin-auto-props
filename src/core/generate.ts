export function generatePropsDefinition(
  compVariable: string,
  props: Record<string, any>,
) {
  return `Object.defineProperty(${compVariable}, "props", ${JSON.stringify({ value: props }, null, 2).replaceAll('"', "")});\n`
}
