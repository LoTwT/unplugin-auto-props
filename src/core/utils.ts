import type { PropertyMeta } from "vue-component-meta"
import type { MappedRuntimeProp } from "./types"

export function mapRuntimeProp(prop: PropertyMeta): MappedRuntimeProp {
  const type = prop.type.replace(" | undefined", "")

  return {
    type: `${type[0].toUpperCase()}${type.slice(1)}`,
    required: prop.required,
  }
}
