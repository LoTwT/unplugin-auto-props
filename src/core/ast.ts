import type { SgNode } from "@ast-grep/napi"
import type { Kinds, TypesMap } from "@ast-grep/napi/types/staticTypes"
import { Lang, parseAsync } from "@ast-grep/napi"
import { DEFINE_COMPONENT } from "./constants"

export async function parseJavaScript(code: string) {
  const astRoot = (await parseAsync(Lang.JavaScript, code)).root()
  return astRoot
}

export function getDefaultExport(node: SgNode<TypesMap, Kinds<TypesMap>>) {
  const defaultExportNode = node.find({
    rule: {
      pattern: "export default $DEFAULT_EXPORT",
      has: {
        field: "value",
        kind: "identifier",
      },
    },
  })
  const defaultExportVariableNode =
    defaultExportNode?.getMatch("DEFAULT_EXPORT") || null
  return {
    defaultExportNode,
    defaultExportVariableNode,
  }
}

export function getComponentDefinition(
  node: SgNode<TypesMap, Kinds<TypesMap>>,
  compVariable: string,
) {
  const componentDefinitionNode = node.find(
    `const ${compVariable} = ${DEFINE_COMPONENT}($FUNC,$OPTION)`,
  )

  const componentOptionNode =
    componentDefinitionNode?.getMatch("OPTION") || null

  return {
    componentDefinitionNode,
    componentOptionNode,
  }
}

export function getPair(
  node: SgNode<TypesMap, Kinds<TypesMap>>,
  keyRegex: string,
) {
  const pairNode = node.find({
    rule: {
      kind: "pair",
      all: [
        {
          has: {
            field: "key",
            regex: keyRegex,
          },
        },
        {
          has: {
            field: "value",
            pattern: "$VALUE",
          },
        },
      ],
    },
  })
  const pairValueNode = pairNode?.getMatch("VALUE") || null

  return {
    pairNode,
    pairValueNode,
  }
}
