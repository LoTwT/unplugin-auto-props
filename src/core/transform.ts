import type { MetaChecker } from "./types"
import MagicString from "magic-string"
import {
  getComponentDefinition,
  getDefaultExport,
  getPair,
  parseJavaScript,
} from "./ast"
import { COMPONENT_OPTION_KEYS, EXTERNAL_META_PROPS } from "./constants"
import { generatePropsDefinition } from "./generate"
import { mapRuntimeProp } from "./utils"

export async function transformDefineComponent(
  code: string,
  id: string,
  checker: MetaChecker,
): Promise<MagicString | null> {
  const astRoot = await parseJavaScript(code)

  const { defaultExportNode, defaultExportVariableNode } =
    getDefaultExport(astRoot)

  if (!defaultExportNode || !defaultExportVariableNode) return null

  const defaultExportVariable = defaultExportVariableNode.text()

  const { componentDefinitionNode, componentOptionNode } =
    getComponentDefinition(astRoot, defaultExportVariable)

  if (componentDefinitionNode && componentOptionNode) {
    const { pairNode: propsNode } = getPair(
      componentOptionNode,
      COMPONENT_OPTION_KEYS.PROPS,
    )

    // TODO: incrementally update props
    if (propsNode) return null
  }

  const meta = checker.getComponentMeta(id)

  const props = meta.props
    .filter((prop) => !EXTERNAL_META_PROPS.includes(prop.name))
    .reduce<Record<string, any>>((res, prop) => {
      res[prop.name] = mapRuntimeProp(prop)
      return res
    }, {})

  const propsDefinition = generatePropsDefinition(defaultExportVariable, props)

  const s = new MagicString(code)

  s.appendLeft(defaultExportNode.range().start.index, propsDefinition)

  return s
}
