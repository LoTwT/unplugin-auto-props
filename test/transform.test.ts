import type { MetaChecker } from "../src/core/types"
import { resolve } from "node:path"
import { fileURLToPath } from "node:url"
import { createChecker } from "vue-component-meta"
import { transformDefineComponent } from "../src/core/transform"
import { getTransformCode } from "./helpers/get-transform-code"

describe("transform", () => {
  const _dirname = fileURLToPath(new URL(".", import.meta.url))

  let checker: MetaChecker
  let codesTransformed: Awaited<ReturnType<typeof getTransformCode>>
  let codesNotTransformed: Awaited<ReturnType<typeof getTransformCode>>

  beforeAll(async () => {
    const tsconfigPath = resolve(_dirname, "./__fixtures__/tsconfig.json")
    checker = createChecker(tsconfigPath, {})

    const idsShouldTransformed = ["base.tsx"].map((id) =>
      resolve(_dirname, "__fixtures__", id),
    )

    const idsShouldNotTransformed = ["base-has-options.tsx"].map((id) =>
      resolve(_dirname, "__fixtures__", id),
    )

    const transformResults = await Promise.all([
      getTransformCode(idsShouldTransformed),
      getTransformCode(idsShouldNotTransformed),
    ])

    codesTransformed = transformResults[0]

    codesNotTransformed = transformResults[1]
  })

  describe("transformDefineComponent", () => {
    it("should transform defineComponent", async () => {
      const codes = await Promise.all(
        Object.entries(codesTransformed).map(async ([id, code]) => {
          const s = await transformDefineComponent(code, id, checker)
          return s?.toString()
        }),
      )

      expect(codes.map((c) => c?.includes("Object.defineProperty"))).toEqual([
        true,
      ])

      expect(codes).toMatchSnapshot()
    })

    it("should not transform defineComponent with options", async () => {
      const codes = await Promise.all([
        transformDefineComponent("", "virtual-no-export-default.ts", checker),
        ...Object.entries(codesNotTransformed).map(async ([id, code]) => {
          const s = await transformDefineComponent(code, id, checker)
          return s?.toString()
        }),
      ])

      expect(codes).toEqual([null, undefined])
    })
  })
})
