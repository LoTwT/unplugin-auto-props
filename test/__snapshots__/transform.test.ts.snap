// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > transformDefineComponent > should transform defineComponent 1`] = `
[
  "import { defineComponent, createTextVNode as _createTextVNode, createVNode as _createVNode } from "vue";
const Foo = defineComponent(props => () => _createVNode("div", null, [_createTextVNode("Foo: "), props.foo, _createTextVNode(" - "), props.disabled?.toString(), _createTextVNode(" - "), typeof props.disabled]));
Object.defineProperty(Foo, "props", {
  value: {
    foo: {
      type: String,
      required: true
    },
    disabled: {
      type: Boolean,
      required: false
    }
  }
});
export default Foo;",
]
`;
